pkg/backup/backup.go

```diff
@@ -264,30 +274,222 @@ func (kb *kubernetesBackupper) Backup(log logrus.FieldLogger, backupRequest *Req
+	if err != nil {
+		return errors.Wrap(err, "error creating temp dir for backup")
+	}
+	defer os.RemoveAll(tempDir)
+
+	collector := &itemCollector{
+		log:                   log,
+		backupRequest:         backupRequest,
+		discoveryHelper:       kb.discoveryHelper,
+		dynamicFactory:        kb.dynamicFactory,
+		cohabitatingResources: cohabitatingResources(),
+		dir:                   tempDir,
+	}
+
+	items := collector.getAllItems()
+	log.WithField("progress", "").Infof("Collected %d items matching the backup spec from the Kubernetes API (actual number of items backed up may be more or less depending on velero.io/exclude-from-backup annotation, plugins returning additional related items to back up, etc.)", len(items))
+
+	patch := fmt.Sprintf(`{"status":{"progress":{"totalItems":%d}}}`, len(items))
+	if _, err := kb.backupClient.Backups(backupRequest.Namespace).Patch(backupRequest.Name, types.MergePatchType, []byte(patch)); err != nil {
+		log.WithError(errors.WithStack((err))).Warn("Got error trying to update backup's status.progress.totalItems")
+	}
 
-	for _, group := range kb.discoveryHelper.Resources() {
-		if err := gb.backupGroup(group); err != nil {
-			log.WithError(err).WithField("apiGroup", group.String()).Error("Error backing up API group")
+	itemBackupper := &itemBackupper{
+		backupRequest:           backupRequest,
+		tarWriter:               tw,
+		dynamicFactory:          kb.dynamicFactory,
+		discoveryHelper:         kb.discoveryHelper,
+		resticBackupper:         resticBackupper,
+		resticSnapshotTracker:   newPVCSnapshotTracker(),
+		volumeSnapshotterGetter: volumeSnapshotterGetter,
+		itemHookHandler: &defaultItemHookHandler{
+			podCommandExecutor: kb.podCommandExecutor,
+		},
+	}
+
+	// helper struct to send current progress between the main
+	// backup loop and the gouroutine that periodically patches
+	// the backup CR with progress updates
+	type progressUpdate struct {
+		totalItems, itemsBackedUp int
+	}
+
+	// the main backup process will send on this channel once
+	// for every item it processes.
+	update := make(chan progressUpdate)
+
+	// the main backup process will send on this channel when
+	// it's done sending progress updates
+	quit := make(chan struct{})
+
+	// This is the progress updater goroutine that receives
+	// progress updates on the 'update' channel. It patches
+	// the backup CR with progress updates at most every second,
+	// but it will not issue a patch if it hasn't received a new
+	// update since the previous patch. This goroutine exits
+	// when it receives on the 'quit' channel.
+	go func() {
+		ticker := time.NewTicker(1 * time.Second)
+		var lastUpdate *progressUpdate
+		for {
+			select {
+			case <-quit:
+				ticker.Stop()
+				return
+			case val := <-update:
+				lastUpdate = &val
+			case <-ticker.C:
+				if lastUpdate != nil {
+					patch := fmt.Sprintf(`{"status":{"progress":{"totalItems":%d,"itemsBackedUp":%d}}}`, lastUpdate.totalItems, lastUpdate.itemsBackedUp)
```

Comment:
Have you tested this against a cluster that had older backup CRs, without the field, installed? What was the failure mode? I think that would be useful to document for folks, since I think a lot of people will just update their image and not read the upgrade doc.

Reply:
I have not tested that but will.

Reply:
I think I made a similar comment :)
How did the testing go?

Reply:
I'll test it out this afternoon and will report back!

Reply:
OK testing that was easier than I thought :) The behavior if the CRDs have not been updated is that the progress updates to the CR are a no-op. I believe this is because `preserveUnknownFields` is false, so the updates to those unknownf fields are pruned by the API server. I think this is the desirable behavior for this feature.

Reply:
Yeah, that is desirable. But we don't want folks not to update their CRDs. I think that can be solved separately.